parquet/src/main/java/org/apache/iceberg/parquet/ParquetAvro.java

```diff
@@ -169,12 +164,14 @@ public String getLogicalTypeName() {
 
     @Override
     public BigDecimal fromFixed(GenericFixed value, Schema schema, LogicalType type) {
-      return super.fromFixed(value, schema, decimalsByScale[((ParquetDecimal) type).scale()]);
+      final ParquetDecimal dec = (ParquetDecimal) type;
+      return super.fromFixed(value, schema, LogicalTypes.decimal(dec.precision(), dec.scale()));
```

Comment:
This change results in creating a new `Decimal` for every value. While I don't think that we use this very much, I would rather not introduce object allocation for every conversion. Can you add a static cache for these? That was the purpose of `decimalsByScale`.

Reply:
I know, but we didn't take the scale into account, so it could actually change the scale. I fixed this in the `fromFixed` by directly creating the `BigDecimal`, this is what's happening in the Avro library:
```
    @Override
    public BigDecimal fromFixed(GenericFixed value, Schema schema, LogicalType type) {
      int scale = ((LogicalTypes.Decimal) type).getScale();
      return new BigDecimal(new BigInteger(value.bytes()), scale);
    }
```
For the `toFixed` this isn't that trivial, since this would mean copying a lot of code.

Reply:
I think it's fine to delegate to `toFixed`. I just don't want to call `LogicalTypes.decimal` every time to create a new decimal `LogicalType` with the same scale and precision. I think just adding a cache keyed by a `Pair` of scale and precision would fix it.